### ISSUE TITLE
Type Guards for union of dictionaries (proper type recognition)

### ISSUE THREAD

**Issue by fider-apptimia** [microsoft/TypeScript#38267]
## Search Terms
- union of dictionaries
- union of objects

## Suggestion

Type recognition for union of multiple dictionaries.

## Use Cases
Type error
```typescript
type A = {
  a: string,
  x: number
};

type B = {
  b: string,
  y: number
};

function fun(ab: A | B): number {
  if (ab.a) { // Property 'a' does not exist on type 'A | B'
    return ab.x // Property 'x' does not exist on type 'B'.
  } else {
    return ab.y // Property 'y' does not exist on type 'A'.
  }
}
```

## Examples

Example above should not throw compilation error.

**Comment by luisgurmendezMLabs** [comment]
You can use type guards for this.
Define a function
```
function isA(a: A | B): a is A {
   return (a as A).a !== undefined
}
```
And the use it like:

```
function fun(ab: A | B): number {
  if (isA(ab)) { // Property 'a' does not exist on type 'A | B'
    return ab.x // Property 'x' does not exist on type 'B'.
  } else {
    return ab.y // Property 'y' does not exist on type 'A'.
  }
}
```

**Comment by nmain** [comment]
Typescript supports this right now when using `in` checks:

[Playground Link](https://www.typescriptlang.org/play/#code/C4TwDgpgBAglC8UDeAoKUCGAuKBnYATgJYB2A5gDRpQAeOJArgLYBGEBKAvgNwoqiQoAIQTJqLHPmLkq6EPWZsOPPgDMGJAMbAiAexJR1JABQYJsKAB9hASgWt2Y9EVVRjAIgzuopTCxvIUAD0QVAACgS6kASgUADkGHFQACa6ELhQJLrAUBA0RPhQ+lAC0HFw1kJx1OgEEMAMBAZmAHQ0waERUeyxcTRJqemZ2bn5hcWl8VUt1Jy5ADa40KjotfWNzSwtIB3hkdG9IANpGVk5eQU5E+BlMHEz6JxcQA)

**Comment by fider-apptimia** [comment]
Thank you for quick response.
